### ISSUE TITLE
FTPClient should support renaming files.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3335]
|<img alt="tarjei's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| tarjei reported|
|-|-|
|Trac ID|trac#3335|
|Type|enhancement|
|Created|2008-07-07 07:06:31Z|
|Branch|https://github.com/twisted/twisted/tree/ftp-rename-3335|

The attached diff provides an implementation of a new rename method on the FTPClient that implements the RNFR/RNTO ftp commands. 

It contains a new method on the ftpclient:
```
    def rename(self, pathFrom, pathTo):
        """
        Rename a file.

        This method issues the 'RNFR, RNTO' command sequence.
        @param: pathFrom: the absolute path to the file to be renamed
        @param: pathTo: the absolute path to rename the file to.
        @return: A list of L{deferrend}s. Add your callbacks to the last one. 
        """

        cmds = ['RNFR ' + self.escapePath(pathFrom),
                'RNTO ' + self.escapePath(pathTo)]
        d1 = self.queueStringCommand(cmds[0])
        d2 = self.queueStringCommand(cmds[1])

        def RNFR_pass(result):
            return d2
        d1.addCallbacks(RNFR_pass, self.fail)
        return d2
```

And a corresponding test:
```
    def testRenameFromTo(self):
        """
        Test renaming a file.

        L{ftp.FTPClient.rename} should return a deferred which fires after a successfull
        RNFR, RNTO sequece to the server.
        """
        self._testLogin()
        def cbFinish(result):
            self.assertEquals(self.transport.value(), 'RNTO /ham\r\n')
        d2 = self.client.rename("/spam", "/ham")
        d2.addCallback(cbFinish)
        self.assertEquals(self.transport.value(), 'RNFR /spam\r\n')
        self.transport.clear()
        self.client.lineReceived('350 Requested file action pending further information.\r\n')
        self.client.lineReceived('250 Requested File Action Completed OK')
        return d2
```

Please review and commit.

Attachments:

* [twisted_ftp_rename_patch.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/90c/90c8af8db7f253890139189d6e908e86ba25a676/9b3d0544af68c871dd8201aee6bd6360566b8d31.py) (2404 bytes) - added by tarjei on 2008-07-07 07:07:35Z - Patch adding rename() to FTPClient
* [twisted_ftp_rename_patch_v2.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/90c/90c8af8db7f253890139189d6e908e86ba25a676/534210302402526e6404a9482a04ebe0318ba445.py) (3863 bytes) - added by tarjei on 2008-07-07 08:26:53Z - Updated patch with 2 new tests
* [twisted_ftp_rename_patch_v3.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/90c/90c8af8db7f253890139189d6e908e86ba25a676/5d9e24ca2f3f33db6a5ac0d7f904f661edbce19b.py) (5420 bytes) - added by tarjei on 2008-07-07 11:02:52Z - Updated patch with make directory method
* [twisted_ftp_rename_patch_v4.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/90c/90c8af8db7f253890139189d6e908e86ba25a676/bdb30b9d0be8a1a2a6a3e81c341958d12a2d4fcd.py) (6353 bytes) - added by tarjei on 2008-07-07 11:28:23Z - Updated version with better parsing of makeDirectory return values.
* [ftp_rename_patch_v5.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/90c/90c8af8db7f253890139189d6e908e86ba25a676/732f1d494983b0a4df66a8baec6f1bbcd5276c10.diff) (6982 bytes) - added by tarjei on 2008-10-24 10:29:17Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3335 3335
type__enhancement enhancement
reporter__tarjei tarjei
priority__normal normal
milestone__ 
branch__branches_ftp_rename_3335 branches/ftp-rename-3335
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ftp__client ftp, client
time__1215414391000000 1215414391000000
changetime__1225315812000000 1225315812000000
version__None None
owner__ 
cc__spiv cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|spiv set owner to tarjei|
|-|-|

This patch looks like a great start.  It's nice to receive a patch with docstrings and a test.

Still, I'd like to see some tests for error handling.  In particular, a test for how an error response for the RNFR command is handled (error is passed to caller via the deferred, RNTO call is cancelled?), and a corresponding test for RNTO.

**Comment by twisted-trac** [comment]
|<img alt="tarjei's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|tarjei commented|
|-|-|

Thank you for the comments. I updatet the patch with the following tests:

```

    def testRenameFromToFailingOnFirstError(self):
        """
        Test renaming a file that does not exist.
        """
        self._testLogin()
        d2 = self.client.rename("/spam", "/ham")
        self.assertEquals(self.transport.value(), 'RNFR /spam\r\n')
        self.transport.clear()
        self.client.lineReceived('550 Requested file unavailable.\r\n')
        # the client disconnects on first error and thus clears the action queue
        self.assertEquals(self.client.actionQueue, [])
        self.assertFailure(d2, ftp.CommandFailed)
        return d2

    def testRenameFromToFailingOnRenameTo(self):
        """
        Test renaming a file to an impossible path
        """
        self._testLogin()
        d2 = self.client.rename("/spam", "/ham")
        self.assertEquals(self.transport.value(), 'RNFR /spam\r\n')
        def assertFail(result):
            self.assertTrue(isinstance(result, failure.Failure))
        d2.addErrback(assertFail)
        self.transport.clear()
        self.client.lineReceived('350 Requested file action pending further information.\r\n')
        self.assertEquals(self.transport.value(), 'RNTO /ham\r\n')
        self.client.lineReceived('550 Requested file unavailable.\r\n')
        self.assertEquals(self.client.actionQueue, [])
        return d2


```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv set owner to spiv|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to tarjei|
|-|-|

Here we go:
 * it would have been easier to review MKD in a separate ticket
 * you should use assertFailure instead of adding errback in tests
 * I suspect rename should return d1 instead of d2. Having better names for these variables would be great
 * the "raise `ValueErrorè" in makeDirectory is not tested. It should be "raise ValueError()", too
 * the escapePath calls are not tested
 * use assertIsInstance instead of assert(isinstance())
 * use assertEqual consistently instead of assertEquals
 * the code should fit in 80 columns
 * testRename* methods should be named test_rename*, and have 2 blank lines as separation
 
Thanks!

**Comment by twisted-trac** [comment]
|<img alt="tarjei's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|tarjei commented|
|-|-|

Just a few notes to myself: 
* I updated the documentation on quit() to say that you have to end the deferred before you call loseConnection()

**Comment by twisted-trac** [comment]
|<img alt="tarjei's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|tarjei commented|
|-|-|

Ok, an extra patch has been uploaded that fixes spivs points. Three clarifications: 

* the escapePath calls are not tested 
-> escapePath is just .replace("\n", "\0") + it is used by all the commmands in that class. I'm not sure what you want to be tested in that function.

* use assertEqual consistently instead of assertEquals 
-> I tried to do a global search and replace on this one, but it changes the patch size so much I didn't thing it is worthwhile.Maybe this should be done globaly?

* I've added an extra documentation comment in the top of the file. This is realy [#3490](https://github.com/twisted/twisted/issues/3490) but I forgot to edit it out before submitting the patch. 

Apart from that, I think this patch fixes the other points.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [25082]) FTPClient.rename

refs [#3335](https://github.com/twisted/twisted/issues/3335)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

  1. There's some trailing whitespace in the patch which I removed.
  1. The types of the parameters to `rename` weren't documented so I documented them.
  1. The docs for the return type of `rename` were out of date (looks like they documented a previous version of the patch), so I updated them.  I also think the returned Deferred should make the response to both commands available, so I change the code to make that happen.
  1. The docstring for `test_renameFromTo` had some typos and wasn't extremely descriptive, so I fixed them and added some more text.
  1. The `makeDirectory` change should be separate, so I split it out.
  1. I changed the new tests to use `assertEqual` instead of `assertEquals`
  1. I added a test for `escapePath` usage.
  1. I backed out the change to the FTPClient docstring, since that's for [#3490](https://github.com/twisted/twisted/issues/3490).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [25086]) apply the makeDirectory parts of ftp_rename_patch_v5.diff

refs [#3335](https://github.com/twisted/twisted/issues/3335)
refs [#3500](https://github.com/twisted/twisted/issues/3500)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra set owner to @exarkun|
|-|-|

1.the @since tag is missing
 1. Failure is missing epytext markup
 1. QUIT is missing epytext markup

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [25102]) improve docstring markup

refs [#3335](https://github.com/twisted/twisted/issues/3335)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun removed owner|
|-|-|

Not sure which `Failure` you meant for 2.  Fixed the other two points.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_16"></a>|@radix commented|
|-|-|

Looks good +1

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set status to `closed`|
|-|-|

(In [25188]) Merge ftp-rename-3335

Author: tarjei, exarkun
Reviewer: spiv, therve, thijs, radix
Fixes: [#3335](https://github.com/twisted/twisted/issues/3335)

Add `FTPClient.rename` method for using the rename feature of FTP.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
